all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,526 @@
+      those that need customized payload conversion will need to create the remote codec themselves.
+* Options like grpc-meta sets header provider, API key sets credentials, etc.
+  * 💭 Why make a user that just wants to _add_ a header on top of config have to use `LoadConfig` and do it the hard
+    way?
+    * There's no clean way to support single header adding on top of existing ones in the Go SDK today.
+
+Simplest example:
+
+```go
+options, err := envconfig.LoadClientOptionsFromConfig("default", LoadClientConfigOptions{})
+if err != nil {
+  return err
+}
+cl, err := client.Dial(options)
+```
+
+#### Java SDK Idea
+
+* New JAR project `temporal-envconfig`.
+  * 💭 Why not just in the SDK?
+    * We have to take a TOML dependency which is not fair to put on users that don't need this functionality.
+* New models for `ClientConfig` (and other things as needed).
+* New static `ClientConfig.load()` and
+  `ClientConfig.load(@Nullable String configFile, boolean disableFile, boolean disableEnv)` that return `ClientConfig`.
+  * ❓Any better place to put this? It's a config utility, not a service stub utility.
+* New instance `ClientConfig#toWorkflowServiceStubsOptions()` and static helpers like
+  `ClientConfig.loadWorkflowServiceStubsOptions()` and the overload like `load` has.
+* Need the same for operator and cloud service stubs options.
+* Need same methods for `WorkflowClientOptions`.
+  * Unfortunately with how Java works, there are separate client options, and they too can be in config.
+  * ❓ Any alternative suggestions here? We could have some kind of shortcut to do multiple steps here.
+* ❓ Does Java have a concept of a remote codec?
+
+Simplest example:
+
+```java
+var stubsOptions = ClientConfig.loadWorkflowServiceStubsOptions();
+var stubs = WorkflowServiceStubs.newServiceStubs(stubsOptions);
+var clientOptions = ClientConfig.loadWorkflowClientOptions();
+var client = WorkflowClient.newInstance(stubs, clientOptions);
+```
+
+#### TypeScript SDK Idea
+
+* New package and module for `envconfig`.
+  * 💭 Why not just in the SDK?
+    * We have to take a TOML dependency which is not fair to put on users that don't need this functionality.
+* Needs `ClientConfig` interface (and other interfaces as needed).
+* Needs `loadClientConfig(options?)` where the `options` type is
+  `{ configFile?: string, disableFile?: bool, disableEnv?: bool }` and it returns a `ClientConfig`.
+* Needs `loadClientOptionsFromConfig(options?)` where the optional `options` are
+  either `ClientConfig` or `{ profile?: string, configFile?: string, disableFile?: bool, disableEnv?: bool }` and
+  it returns a `ConnectionOptions & ClientOptions`.
+  * Could also have it return `{ connectionOptions: ConnectionOptions, clientOptions: ClientOptions }` if that makes
+    more sense than a union.
+* TypeScript SDK _does not_ support codec settings and will error if seen in config.
```

Comment:
> Because a user with a profile configured with a codec should expect that codec to be used, not silently ignored.

I agree with your argument, but I'm not totally comfortable with the consequence.

To error in such cases implies that a user who has codecs AND is using an SDK that doesn't have remote codec support would have to create two distinct profiles, one for use with CLI with remote codec configured, and the other for use with his SDK, without remote codec configured. That's very inconvenient.

Not saying that's wrong, but to be coherent with this decision, we will need to move forward with implementing remote codecs in all SDKs. Not difficult, but IIRC, there were doubts about either that's a good idea.

Otherwise, we may find ourselves needing to add an extra flag to `LoadClientOptions` in all SDKs that don't have support for remote codecs, to intentionally ignore remote codec settings from config file/environment (i.e. just turn off that specific error). That's ugly. I'd definitely rather not get down to this.

Reply:
:+1: I agree we'd need to support remote codecs in each SDK to get the benefit of this setting in every SDK. In the meantime, erroring is the clearest approach I think.